### ISSUE TITLE
You do not have sufficient permissions to access this page...

### ISSUE THREAD

**Issue by reholmes27** [engintron/engintron#1156]
I have installed engintron as root, via SSH, but when I click on the plugin page in WHM I get the message in the title. I'm logged in as a reseller that has all privileges and it all appears as if Nginx is running, I just can't access the admin page?

Thanks

Rich

**Comment by etechwebcomm** [comment]
This has been reported a few times - there are specific calls cPanel has for integrating into the reseller center ACL list which must be included in the third party plugin. I know we have a number of other software extensions working great with reseller permissions, so hopefully this can be made to work in Engintron as well!

**Comment by fevangelou** [comment]
There is already a check for resellers as you can see here: https://github.com/engintron/engintron/blob/master/app/engintron.php#L21

**Comment by etechwebcomm** [comment]
@fevangelou Thank you for the response! I believe there must be a registration script to make the software active in the actual ACL. Right now Engintron is not even a selectable element in the WHM "Edit Reseller Nameservers and Privileges" UI. Im trying to pull apart the installation scripts for CSF (configserver security firewall) to see how they implement this, but since Im not a plugin dev Im not 100% certain specifically what to look for.

**Comment by fevangelou** [comment]
Engintron checks which resellers are set to "all" permissions. If you're not, you won't have access to it.

**Comment by etechwebcomm** [comment]
Can confirm "You do not have sufficient permissions to access this page..." error occurs on the reseller accounts we have set to "all". Whenever we manage Engintron, we always have to go through the root administrator.

**Comment by fevangelou** [comment]
Please double check that when permissions are set to "all" that the word "all" is also included in the file: /var/cpanel/resellers

**Comment by etechwebcomm** [comment]
Hi @fevangelou - confirmed. 'all' is included:

`reselleraccount:acct-summary,all,allow-shell,basic-system-info,basic-whm-functions,cors-proxy-get,cpanel-api,cpanel-integration,create-user-session,digest-auth,generate-email-config,list-pkgs,manage-api-tokens,manage-dns-records,manage-oidc,manage-styles,mysql-info,ns-config,public-contact,ssl-info,track-email`

**Comment by fevangelou** [comment]
The check works then. You should be able to access it.

**Comment by etechwebcomm** [comment]
Can confirm "You do not have sufficient permissions to access this page..." 
Tested on 4 "all" enabled resellers on two different server hosts.

**Comment by reholmes27** [comment]
I have managed to make some progress. The checkacl() function needs to have str_split replaced with explode:

`if (file_exists('/var/cpanel/resellers') && is_readable('/var/cpanel/resellers')) {
        $reseller = file_get_contents('/var/cpanel/resellers');
        if (trim($reseller) != '') {
            foreach (explode("\n", $reseller) as $line) {
                if (preg_match("/^$user:/", $line)) {
                    $line = preg_replace("/^$user:/", "", $line);
                    foreach (explode(",", $line) as $perm) {
                        if ($perm == "all") {
                            return 1;
                        }
                    }
                }
            }
        }
    }`

This then accepts the permission but causes a 500 error.  This is because a call to file_get_contents(/var/cpanel/caches/_generated_header_files/cache_keys.json); returns, 'Permission Denied'. I've tried setting that file to 777 reseller:reseller but still the same.

**Comment by etechwebcomm** [comment]
@fevangelou - the inserted code you mentioned to engintron.php results in the following:
`int(0) You do not have sufficient permissions to access this page...`

**Comment by reholmes27** [comment]
> @fevangelou - the inserted code you mentioned to engintron.php results in the following:
> `int(0) You do not have sufficient permissions to access this page...`

You get this if you don't make the change I've shown above.

**Comment by fevangelou** [comment]
@etechwebcomm Well, @reholmes27 is right. str_split() is wrongly used in its first instance. I've updated the code to use better checks for newline characters so it's always split into proper lines.

Update Engintron to get the changed code and let me know.

**Comment by etechwebcomm** [comment]
Ok, we manually updated from mainline - unfortunately reseller access to Engintron now errors out with an error 500.

**Comment by reholmes27** [comment]
The 500 error is because of this:

`Warning: file_get_contents(/var/cpanel/caches/_generated_header_files/cache_keys.json): failed to open stream: Permission denied in /usr/local/cpanel/php/WHM.php on line 61

Fatal error: Uncaught Exception: User does not exist in cache_keys. Try regenerating the cache. in /usr/local/cpanel/php/WHM.php:66 Stack trace: #0 /usr/local/cpanel/php/WHM.php(47): WHM::getUserCacheKey('ampersand') #1 /usr/local/cpanel/php/WHM.php(10): WHM::processHeader('Engintron for c...', 1, 1) #2 /usr/local/cpanel/whostmgr/docroot/cgi/engintron.php(488): WHM::header('Engintron for c...', 1, 1) #3 {main} thrown in /usr/local/cpanel/php/WHM.php on line 66`

Basically access to the cache_keys.json file is denied unless root user. I've tried changing owner and permissions of the file but it doesn't make a difference unfortunately.

**Comment by fevangelou** [comment]
This is probably a bug in the PHP API of WHM.
